File: templates/variants/go-connectrpc/internal/connectapi/list_messages_integration_test.go
package connectapi

import (
	"context"
	"fmt"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
	"testing"
	"time"

	"connectrpc.com/connect"
	_ "github.com/jackc/pgx/v5/stdlib"
	"github.com/jmoiron/sqlx"

	chatv1 "{{MODULE_PATH}}/gen/chat/v1"
	chatv1connect "{{MODULE_PATH}}/gen/chat/v1/chatv1connect"
	"{{MODULE_PATH}}/internal/app"
)

func TestListMessagesPaginationIncludesAttachmentMetadata(t *testing.T) {
	databaseURL := strings.TrimSpace(os.Getenv("DATABASE_URL"))
	if databaseURL == "" {
		t.Skip("DATABASE_URL is required for integration test")
	}

	t.Setenv("ATTACHMENT_PUBLIC_BASE_URL", "https://storage.test")

	db, err := app.OpenDatabase(context.Background(), databaseURL)
	if err != nil {
		t.Fatalf("open database: %v", err)
	}
	t.Cleanup(func() { _ = db.Close() })

	if _, err := db.ExecContext(context.Background(), `
truncate table
  webhook_events,
  attachments,
  messages,
  conversation_participants,
  conversations,
  users
restart identity cascade`); err != nil {
		t.Fatalf("truncate tables: %v", err)
	}

	storage := newFakeStorage()
	service := app.NewChatService(db, storage, app.GenericWebhookAdapter{})
	path, handler := NewHandler(service)
	mux := http.NewServeMux()
	mux.Handle(path, handler)
	server := httptest.NewServer(mux)
	t.Cleanup(server.Close)

	client := chatv1connect.NewChatServiceClient(http.DefaultClient, server.URL)

	createUserResponse, err := client.CreateUser(context.Background(), connect.NewRequest(&chatv1.CreateUserRequest{
		Username:    "alice",
		DisplayName: "Alice",
	}))
	if err != nil {
		t.Fatalf("create user: %v", err)
	}
	userID := createUserResponse.Msg.GetUser().GetId()

	createConversationResponse, err := client.CreateConversation(context.Background(), connect.NewRequest(&chatv1.CreateConversationRequest{
		CreatedByUserId:    userID,
		Title:              "General",
		ParticipantUserIds: []string{userID},
	}))
	if err != nil {
		t.Fatalf("create conversation: %v", err)
	}
	conversationID := createConversationResponse.Msg.GetConversation().GetId()

	messageIDs := make([]string, 0, 55)
	for index := 1; index <= 55; index++ {
		response, err := client.CreateMessage(context.Background(), connect.NewRequest(&chatv1.CreateMessageRequest{
			ConversationId: conversationID,
			UserId:         userID,
			Body:           fmt.Sprintf("message-%d", index),
		}))
		if err != nil {
			t.Fatalf("create message %d: %v", index, err)
		}
		messageIDs = append(messageIDs, response.Msg.GetMessage().GetId())
	}
	rewriteMessageTimestamps(t, db, messageIDs)

	uploadResponse, err := client.CreateAttachmentUpload(context.Background(), connect.NewRequest(&chatv1.CreateAttachmentUploadRequest{
		ConversationId: conversationID,
		UserId:         userID,
		Filename:       "photo.png",
		ContentType:    "image/png",
		ByteSize:       1234,
	}))
	if err != nil {
		t.Fatalf("create attachment upload: %v", err)
	}
	storage.set(uploadResponse.Msg.GetAttachment().GetPublicUrl(), "image/png", 1234)
	if _, err := client.FinalizeAttachment(context.Background(), connect.NewRequest(&chatv1.FinalizeAttachmentRequest{
		AttachmentId: uploadResponse.Msg.GetAttachment().GetId(),
		MessageId:    messageIDs[54],
	})); err != nil {
		t.Fatalf("finalize attachment: %v", err)
	}

	firstPage, err := client.ListMessages(context.Background(), connect.NewRequest(&chatv1.ListMessagesRequest{
		ConversationId: conversationID,
	}))
	if err != nil {
		t.Fatalf("list messages first page: %v", err)
	}
	if len(firstPage.Msg.GetMessages()) != 50 {
		t.Fatalf("expected 50 messages, got %d", len(firstPage.Msg.GetMessages()))
	}
	if firstPage.Msg.GetMessages()[0].GetBody() != "message-55" {
		t.Fatalf("expected newest message first, got %s", firstPage.Msg.GetMessages()[0].GetBody())
	}
	if firstPage.Msg.GetMessages()[49].GetBody() != "message-6" {
		t.Fatalf("expected oldest message on first page to be message-6, got %s", firstPage.Msg.GetMessages()[49].GetBody())
	}
	if firstPage.Msg.GetNextCursor() == "" {
		t.Fatal("expected next cursor on first page")
	}
	attachments := firstPage.Msg.GetMessages()[0].GetAttachments()
	if len(attachments) != 1 {
		t.Fatalf("expected attachment metadata on newest message, got %#v", attachments)
	}
	if attachments[0].GetPublicUrl() != uploadResponse.Msg.GetAttachment().GetPublicUrl() {
		t.Fatalf("expected public_url %s, got %s", uploadResponse.Msg.GetAttachment().GetPublicUrl(), attachments[0].GetPublicUrl())
	}

	secondPage, err := client.ListMessages(context.Background(), connect.NewRequest(&chatv1.ListMessagesRequest{
		ConversationId: conversationID,
		Cursor:         firstPage.Msg.GetNextCursor(),
	}))
	if err != nil {
		t.Fatalf("list messages second page: %v", err)
	}
	expectedBodies := []string{"message-5", "message-4", "message-3", "message-2", "message-1"}
	if len(secondPage.Msg.GetMessages()) != len(expectedBodies) {
		t.Fatalf("expected %d messages on second page, got %d", len(expectedBodies), len(secondPage.Msg.GetMessages()))
	}
	for index, body := range expectedBodies {
		if secondPage.Msg.GetMessages()[index].GetBody() != body {
			t.Fatalf("expected %s at index %d, got %s", body, index, secondPage.Msg.GetMessages()[index].GetBody())
		}
	}
	if secondPage.Msg.GetNextCursor() != "" {
		t.Fatalf("expected no next cursor on final page, got %s", secondPage.Msg.GetNextCursor())
	}
}

type fakeStorage struct {
	metadata map[string]struct {
		contentType string
		byteSize    int64
		publicURL   string
	}
}

func newFakeStorage() *fakeStorage {
	return &fakeStorage{metadata: map[string]struct {
		contentType string
		byteSize    int64
		publicURL   string
	}{}}
}

func (f *fakeStorage) CreateSignedUpload(_ context.Context, attachmentID string, conversationID string, filename string, contentType string) (string, string, app.UploadTarget, string, error) {
	key := fmt.Sprintf("attachments/%s/%s/%s", conversationID, attachmentID, filename)
	return "test-bucket", key, app.UploadTarget{
		Method: http.MethodPut,
		URL:    "https://uploads.test/" + key,
		Headers: map[string]string{
			"Content-Type": contentType,
		},
	}, "https://storage.test/" + key, nil
}

func (f *fakeStorage) GetObjectMetadata(_ context.Context, bucket string, key string) (string, int64, string, error) {
	entry, ok := f.metadata[bucket+"/"+key]
	if !ok {
		return "", 0, "", fmt.Errorf("missing metadata for %s/%s", bucket, key)
	}
	return entry.contentType, entry.byteSize, entry.publicURL, nil
}

func (f *fakeStorage) set(publicURL string, contentType string, byteSize int64) {
	key := strings.TrimPrefix(publicURL, "https://storage.test/")
	f.metadata["test-bucket/"+key] = struct {
		contentType string
		byteSize    int64
		publicURL   string
	}{
		contentType: contentType,
		byteSize:    byteSize,
		publicURL:   publicURL,
	}
}

func rewriteMessageTimestamps(t *testing.T, db *sqlx.DB, messageIDs []string) {
	t.Helper()
	baseTime := time.Date(2026, time.January, 1, 0, 0, 0, 0, time.UTC)
	for index, messageID := range messageIDs {
		createdAt := baseTime.Add(time.Duration(index+1) * time.Second)
		if _, err := db.ExecContext(context.Background(), `
update messages
set created_at = $2, updated_at = $2
where id = $1`, messageID, createdAt); err != nil {
			t.Fatalf("rewrite message timestamp: %v", err)
		}
	}
}

File: templates/variants/bun-hono/test/list-messages.integration.test.ts
import { afterEach, beforeEach, expect, test } from "bun:test";
import { SQL } from "bun";
import { createApp } from "../src/index";
import { DefaultChatService } from "../src/chat/service";
import { ChatRepository } from "../src/db/repository";
import { createDb } from "../src/db/client";
import type { AttachmentObjectMetadata, AttachmentUploadTarget } from "../src/chat/types";
import type { AttachmentStorage } from "../src/storage";

const databaseUrl = Bun.env.DATABASE_URL?.trim();
const integrationTest = databaseUrl ? test : test.skip;

let sql: SQL | null = null;

beforeEach(async () => {
  if (!databaseUrl) {
    return;
  }
  sql = new SQL(databaseUrl);
  await sql.unsafe(`
    truncate table
      webhook_events,
      attachments,
      messages,
      conversation_participants,
      conversations,
      users
    restart identity cascade
  `);
});

afterEach(async () => {
  await sql?.end();
  sql = null;
});

integrationTest("list messages returns newest-first pages with attachment metadata", async () => {
  Bun.env.ATTACHMENT_PUBLIC_BASE_URL = "https://storage.test";
  const storage = new FakeAttachmentStorage();
  const app = createApp(new DefaultChatService(new ChatRepository(createDb(databaseUrl)), storage, new NoopWebhookAdapter()));

  const user = await createUser(app);
  const conversation = await createConversation(app, user.id);
  const messageIds: string[] = [];

  for (let index = 1; index <= 55; index += 1) {
    const message = await createMessage(app, conversation.id, user.id, `message-${index}`);
    messageIds.push(message.id);
  }

  await rewriteMessageTimestamps(messageIds);

  const uploadResult = await requestJson(app, "/v1/attachments/uploads", {
    method: "POST",
    body: {
      conversation_id: conversation.id,
      user_id: user.id,
      filename: "photo.png",
      content_type: "image/png",
      byte_size: 1234,
    },
    expectedStatus: 201,
  });
  const attachment = uploadResult.result.attachment as {
    id: string;
    publicUrl: string;
    filename: string;
    contentType: string;
    byteSize: number;
    status: string;
  };

  storage.setObjectMetadataFromUrl(uploadResult.result.attachment.publicUrl, {
    contentType: "image/png",
    byteSize: 1234,
    publicUrl: uploadResult.result.attachment.publicUrl,
  });

  await requestJson(app, `/v1/attachments/${attachment.id}/finalize`, {
    method: "POST",
    body: {
      message_id: messageIds[54],
    },
  });

  const firstPage = await requestJson(app, `/v1/conversations/${conversation.id}/messages`);
  expect(firstPage.messages).toHaveLength(50);
  expect(firstPage.messages[0].body).toBe("message-55");
  expect(firstPage.messages[49].body).toBe("message-6");
  expect(firstPage.next_cursor).toBeString();
  expect(firstPage.messages[0].attachments).toEqual([
    {
      id: attachment.id,
      filename: "photo.png",
      content_type: "image/png",
      byte_size: 1234,
      status: "ready",
      public_url: attachment.publicUrl,
    },
  ]);

  const secondPage = await requestJson(
    app,
    `/v1/conversations/${conversation.id}/messages?cursor=${encodeURIComponent(firstPage.next_cursor)}`
  );
  expect(secondPage.messages.map((message: { body: string }) => message.body)).toEqual([
    "message-5",
    "message-4",
    "message-3",
    "message-2",
    "message-1",
  ]);
  expect(secondPage.next_cursor).toBeUndefined();

  const invalidLimit = await app.request(`/v1/conversations/${conversation.id}/messages?limit=0`);
  expect(invalidLimit.status).toBe(400);
  expect(await invalidLimit.json()).toEqual({
    error: "limit must be a positive integer",
    code: "invalid_limit",
  });

  const tooLargeLimit = await app.request(`/v1/conversations/${conversation.id}/messages?limit=101`);
  expect(tooLargeLimit.status).toBe(400);
  expect(await tooLargeLimit.json()).toEqual({
    error: "limit must be at most 100",
    code: "invalid_limit",
  });
});

async function createUser(app: ReturnType<typeof createApp>) {
  const response = await requestJson(app, "/v1/users", {
    method: "POST",
    body: {
      username: "alice",
      display_name: "Alice",
    },
    expectedStatus: 201,
  });
  return response.user as { id: string };
}

async function createConversation(app: ReturnType<typeof createApp>, createdByUserId: string) {
  const response = await requestJson(app, "/v1/conversations", {
    method: "POST",
    body: {
      created_by_user_id: createdByUserId,
      title: "General",
      participant_user_ids: [createdByUserId],
    },
    expectedStatus: 201,
  });
  return response.conversation as { id: string };
}

async function createMessage(app: ReturnType<typeof createApp>, conversationId: string, userId: string, body: string) {
  const response = await requestJson(app, `/v1/conversations/${conversationId}/messages`, {
    method: "POST",
    body: {
      user_id: userId,
      body,
    },
    expectedStatus: 201,
  });
  return response.message as { id: string };
}

async function requestJson(
  app: ReturnType<typeof createApp>,
  path: string,
  input: {
    method?: string;
    body?: unknown;
    expectedStatus?: number;
  } = {}
) {
  const response = await app.request(path, {
    method: input.method ?? "GET",
    headers: input.body ? { "Content-Type": "application/json" } : undefined,
    body: input.body ? JSON.stringify(input.body) : undefined,
  });
  expect(response.status).toBe(input.expectedStatus ?? 200);
  return response.json();
}

async function rewriteMessageTimestamps(messageIds: string[]) {
  if (!sql) {
    throw new Error("sql client not initialized");
  }

  const baseTime = Date.parse("2026-01-01T00:00:00.000Z");
  for (const [index, messageId] of messageIds.entries()) {
    const createdAt = new Date(baseTime + (index + 1) * 1000).toISOString();
    await sql`
      update messages
      set created_at = ${createdAt}::timestamptz,
          updated_at = ${createdAt}::timestamptz
      where id = ${messageId}
    `;
  }
}

class FakeAttachmentStorage implements AttachmentStorage {
  private readonly metadata = new Map<string, AttachmentObjectMetadata>();

  async createSignedUpload(input: {
    attachmentId: string;
    conversationId: string;
    filename: string;
    contentType: string;
  }): Promise<{ bucket: string; key: string; upload: AttachmentUploadTarget; publicUrl: string }> {
    const bucket = "test-bucket";
    const key = `attachments/${input.conversationId}/${input.attachmentId}/${input.filename}`;
    const publicUrl = `https://storage.test/${key}`;
    return {
      bucket,
      key,
      upload: {
        method: "PUT",
        url: `https://uploads.test/${key}`,
        headers: { "Content-Type": input.contentType },
      },
      publicUrl,
    };
  }

  async getObjectMetadata(input: { bucket: string; key: string }) {
    const metadata = this.metadata.get(`${input.bucket}/${input.key}`);
    if (!metadata) {
      throw new Error(`missing metadata for ${input.bucket}/${input.key}`);
    }
    return metadata;
  }

  setObjectMetadataFromUrl(publicUrl: string, input: Omit<AttachmentObjectMetadata, "bucket" | "key">) {
    const [, bucketAndKey = ""] = publicUrl.split("https://storage.test/");
    this.metadata.set(`test-bucket/${bucketAndKey}`, {
      bucket: "test-bucket",
      key: bucketAndKey,
      contentType: input.contentType,
      byteSize: input.byteSize,
      publicUrl: input.publicUrl,
    });
  }
}

class NoopWebhookAdapter {
  async normalize() {
    return {
      provider: "generic",
      externalEventId: "evt_test",
      eventType: "generic.event",
      signatureValid: true,
      payloadJson: "{}",
    };
  }
}
